packages/ai-jsx/src/batteries/constrained-output.tsx

```diff
@@ -5,169 +5,255 @@
-  error: string;
-}
+export type ObjectCompletion = ModelPropsWithChildren & {
+  /** Validators are used to ensure that the final object looks as expected. */
+  validators?: ((obj: object) => void)[];
+  /**
+   * The object Schema that is required. This is a type of validator with the excpetion
+   * that the schema description is also provided to the model.
+   */
+  schema?: z.Schema;
+  /** Any output example to be shown to the model. */
+  example?: string;
+  // TODO (@farzad): better name/framing for example.
+};
+
+export type TypedObjectCompletion = ObjectCompletion & {
+  /** Human-readable name of the type, e.g. JSON or YAML. */
+  typeName: string;
+  /**
+   * Object parser: creates the object to be evaluated given the string.
+   * Note: the parser is only used for validation. The final output is still a string.
+   */
+  parser: (str: string) => object;
+  /**
+   * The intermediate results that get yielded might be incomplete and as such
+   * a cleaning/healing/untruncating step might be required.
+   * For example, see the `untrunctate-json` package.
+   */
+  partialResultCleaner?: (str: string) => string;
+};
+
+export type TypedObjectCompletionWithRetry = TypedObjectCompletion & { retries?: number };
 
-// TODO: schema
 /**
  * A {@link ChatCompletion} component that constrains the output to be a valid JSON string.
  * It uses a combination of prompt engineering and validation with retries to ensure that the output is valid.
  *
+ * Though not required, you can provide a Zod schema to validate the output against. This is useful for
+ * ensuring that the output is of the expected type.
+ *
  * @example
  * ```tsx
+ * const FamilyTree: z.Schema = z.array(
+ *   z.object({
+ *     name: z.string(),
+ *     children: z.lazy(() => FamilyTree).optional(),
+ *   })
+ * );
+ *
+ * return (
  *    <JsonChatCompletion>
  *     <UserMessage>
  *      Create a nested family tree with names and ages.
  *      It should include a total of 5 people.
  *     </UserMessage>
  *    </JsonChatCompletion>
+ * );
  * ```
- *
- * @param retries The maximum number of times to retry the completion if the output is invalid.
- * @param children The children to render.
  * @returns A string that is a valid JSON or throws an error after `retries` attempts.
+ *    Intermediate results that are valid, are also yielded.
  */
-export function JsonChatCompletion({ children, ...props }: { children: AI.Node }) {
-  return (
-    <ObjectFormatChatCompletion typeName="JSON" validator={isJsonString} {...{ ...props }}>
-      {children}
-    </ObjectFormatChatCompletion>
+export async function* JsonChatCompletion(
+  props: Omit<TypedObjectCompletionWithRetry, 'typeName' | 'parser' | 'partialResultCleaner'>,
+  { render }: AI.ComponentContext
+) {
+  return yield* render(
+    <ObjectCOmpletionWithRetry
+      {...props}
+      typeName="JSON"
+      parser={JSON.parse}
+      // TODO: can we remove .default?
+      partialResultCleaner={untruncateJson.default}
```

Comment:
Not sure why I had to do this. I tried different variations of the `import` but couldn't make it work right.

Reply:
Ugh. ESM stuff can be a huge pain.